tests/unittests/Foundation/NSUndoManagerTests.mm

```diff
@@ -0,0 +1,718 @@
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+
+    [object updateNumber:newValue];
+
+    EXPECT_EQ([object.undoManager canUndo], YES);
+    EXPECT_EQ(object.value.intValue, newValue.intValue);
+
+    [object.undoManager endUndoGrouping];
+    [object.undoManager undo];
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+    EXPECT_EQ([object.undoManager canUndo], NO);
+}
+
+TEST(NSUndoManager, NestedUndoGroupsAndActionsNoAuto) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+    NSNumber* val5 = [NSNumber numberWithInt:9];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:val1] autorelease];
+    object.undoManager.groupsByEvent = NO;
+
+    [object.undoManager beginUndoGrouping];
+    [object updateNumber:val2];
+    [object.undoManager endUndoGrouping];
+
+    EXPECT_EQ(object.value.intValue, val2.intValue);
+
+    [object.undoManager beginUndoGrouping];
+    [object.undoManager beginUndoGrouping];
+    [object updateNumber:val3];
+    [object.undoManager endUndoGrouping];
+    [object updateNumber:val4];
+    [object.undoManager endUndoGrouping];
+
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+
+    [object.undoManager undo];
+
+    EXPECT_EQ(object.value.intValue, val2.intValue);
+    [object.undoManager undo];
+    EXPECT_EQ(object.value.intValue, val1.intValue);
+}
+
+TEST(NSUndoManager, NestedUndoNestedGroupsAndActionsNoAuto) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+    NSNumber* val5 = [NSNumber numberWithInt:9];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:val1] autorelease];
+    object.undoManager.groupsByEvent = NO;
+
+    [object.undoManager beginUndoGrouping];
+    [object updateNumber:val2];
+    [object.undoManager endUndoGrouping];
+
+    [object.undoManager beginUndoGrouping];
+    [object updateNumber:val3];
+    [object.undoManager endUndoGrouping];
+
+    [object.undoManager beginUndoGrouping];
+    [object updateNumber:val4];
+    [object.undoManager endUndoGrouping];
+
+    [object.undoManager undoNestedGroup];
+    EXPECT_EQ(object.value.intValue, val3.intValue);
+
+    [object.undoManager undoNestedGroup];
+    EXPECT_EQ(object.value.intValue, val2.intValue);
+}
+
+// Undo test object that references an externally owned undo manager.
+@interface UndoTestObjectWithSharedManager : NSObject
+@property (retain) NSUndoManager* undoManager;
+@property (retain) NSNumber* value;
+- (id)initWithValue:(NSNumber*)num;
+- (void)updateTarget:(NSNumber*)value;
+@end
+
+@implementation UndoTestObjectWithSharedManager
+- (id)initWithValue:(NSNumber*)num undoManager:(NSUndoManager*)manager {
+    self = [super init];
+    _undoManager = manager;
```

Comment:
Heck, it says it's a `retain` property.

---
In reply to: [77098028](https://github.com/Microsoft/WinObjC/pull/900#discussion_r77098028) [](ancestors = 77098028)